libs/constrained-generators/src/Constrained/GenT.hs

```diff
@@ -224,6 +226,20 @@ suchThatT g p = do
     go n cont = do
       a <- g
       if p a then pure a else scaleT (+ 1) $ go (n - 1) cont
+-}
+
+suchThatWithTryT :: MonadGenError m => Int -> GenT m a -> (a -> Bool) -> GenT m a
+suchThatWithTryT trys g p = do
+  mode <- getMode
+  let (n, cont) = case mode of
+        Strict -> (trys, fatalError)
+        Loose -> (1 :: Int, genError) -- TODO: Maybe 1 is not the right number here!
+  go n cont
+  where
+    go 0 cont = cont (pure ("Ran out of tries (" ++ show trys ++ ") on suchThatWithTryT"))
+    go n cont = do
```

Comment:
If supplied `tries` is negative it could go into infinite loop, if `g` always produces `True` that is.
```suggestion
    go n cont
     | n <= cont (pure ("Ran out of tries (" ++ show trys ++ ") on suchThatWithTryT"))
     | otherwise = do
```

Reply:
That's what I meant to suggest :smile: